base/linalg/lq.jl

```diff
@@ -0,0 +1,145 @@
+# This file is a part of Julia. License is MIT: http://julialang.org/license
+
+# LQ Factorizations
+
+immutable LQ{T,S<:AbstractMatrix} <: Factorization{T}
+    factors::S
+    τ::Vector{T}
+    LQ(factors::AbstractMatrix{T}, τ::Vector{T}) = new(factors, τ)
+end
+
+immutable LQPackedQ{T} <: AbstractMatrix{T}
+    factors::Matrix{T}
+    τ::Vector{T}
+end
+
+LQ{T}(factors::AbstractMatrix{T}, τ::Vector{T}) = LQ{T,typeof(factors)}(factors, τ)
+
+lqfact!{T<:BlasFloat}(A::StridedMatrix{T}) = LQ(LAPACK.gelqf!(A)...)
+lqfact{T<:BlasFloat}(A::StridedMatrix{T})  = lqfact!(copy(A))
+lqfact(x::Number) = lqfact(fill(x,1,1))
+
+function lq(A::Union{Number, AbstractMatrix}; thin::Bool=true)
+    F = lqfact(A)
+    F[:L], full(F[:Q], thin=thin)
+end
+
+convert{T}(::Type{LQ{T}},A::LQ) = LQ(convert(AbstractMatrix{T}, A.factors), convert(Vector{T}, A.τ))
+convert{T}(::Type{Factorization{T}}, A::LQ) = convert(LQ{T}, A)
+
+function getindex(A::LQ, d::Symbol)
+    m, n = size(A)
+    if d == :L
+        return tril!(A.factors[1:m, 1:min(m,n)])
+    elseif d == :Q
+        return LQPackedQ(A.factors,A.τ)
+    else
+        throw(KeyError(d))
+    end
+end
+
+getindex(A::LQPackedQ, i::Integer, j::Integer) = (x = zeros(eltype(A), size(A, 1)); x[i] = 1; y = zeros(eltype(A), size(A, 2)); y[j] = 1; dot(x, A*y))
+
+getq(A::LQ) = LQPackedQ(A.factors, A.tau)
+
+convert{T}(::Type{LQPackedQ{T}}, Q::LQPackedQ) = LQPackedQ(convert(AbstractMatrix{T}, Q.factors), convert(Vector{T}, Q.τ))
+convert{T}(::Type{AbstractMatrix{T}}, Q::LQPackedQ) = convert(LQPackedQ{T}, Q)
+
+size(A::LQ, dim::Integer) = size(A.factors, dim)
+size(A::LQ) = size(A.factors)
+size(A::LQPackedQ, dim::Integer) = 0 < dim ? (dim <= 2 ? size(A.factors, dim) : 1) : throw(BoundsError())
+size(A::LQPackedQ) = size(A.factors)
+
+full(A::LQ) = A[:L]*A[:Q]
+function full{T}(A::LQPackedQ{T}; thin::Bool=true)
+    if thin
+        LAPACK.orglq!(copy(A.factors),A.τ)
+    else
+        A_mul_B!(A, eye(T, size(A.factors,2), size(A.factors,1)))
```

Comment:
Isn't this just an indexing operation? It's not intuitively obvious which argument is modified in 2 arg `A_mul_B!`.

Explicitly constructing a dense `eye` almost always looks like a code smell, lots of wasted memory for zeros you don't care about.